server/svix-server/src/queue/redis.rs

```diff
@@ -114,39 +117,60 @@ async fn background_task(
 
     let mut pool = pool.get().await?;
 
-    // First look for delayed keys whose time is up and add them to the main qunue
-    let timestamp = Utc::now().timestamp();
-    let keys: Vec<String> = pool
-        .zrangebyscore_limit(&delayed_queue_name, 0isize, timestamp, 0isize, batch_size)
-        .await?;
+    // There is a lock on the delayed queue processing to avoid race conditions. So first try to
+    // acquire the lock should it not already exist. The lock expires after five seconds in case a
+    // worker crashes while holding the lock.
+    let mut cmd = redis::cmd("SET");
+    cmd.arg(DELAYED_LOCK)
+        .arg(true)
+        .arg("NX")
+        .arg("PX")
+        .arg(5000);
+    // WIll be Some("OK") when set or None when not set
+    let resp: Option<String> = pool.query_async(cmd).await?;
+
+    if resp.as_deref() == Some("OK") {
+        // First look for delayed keys whose time is up and add them to the main qunue
+        let timestamp = Utc::now().timestamp();
+        let keys: Vec<String> = pool
+            .zrangebyscore_limit(&delayed_queue_name, 0isize, timestamp, 0isize, batch_size)
+            .await?;
 
-    if !keys.is_empty() {
-        // FIXME: needs to be a transaction
-        let keys: Vec<(String, String)> = pool
-            .zpopmin(&delayed_queue_name, keys.len() as isize)
-            .await
-            .unwrap();
-        let tasks: Vec<&str> = keys
-            .iter()
-            // All information is stored in the key in which the ID and JSON formated task
-            // are separated by a `|`. So, take the key, then take the part after the `|`
-            .map(|x| &x.0)
-            .map(|x| x.split('|').nth(1).expect("Improper key format"))
-            .collect();
-
-        // Then for each task, XADD them to ghe MAIN queue
-        let mut pipe = redis::pipe();
-        for task in tasks {
-            let _ = pipe.xadd(
-                &main_queue_name,
-                GENERATE_STREAM_ID,
-                &[(QUEUE_KV_KEY, task)],
-            );
+        if !keys.is_empty() {
+            // Delay every key to be scheduled ten seconds in the future so if the worker crashes
+            // during this block, the messages aren't lost
+            let mut pipe = redis::pipe();
+            for key in &keys {
+                let _ = pipe.zincr(&delayed_queue_name, key, 10_000);
+            }
```

Comment:
There's no need for this anymore because you're locking access to this so if the worker crashes they'll just be picked up again.